File: Medium/0002-add-two-numbers/python/addtwonumber.py
# LeetCode-style definition:
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        # Dummy node to simplify head logic
        dummy = ListNode(0)
        cur = dummy
        carry = 0

        while l1 is not None or l2 is not None or carry != 0:
            s = carry

            if l1 is not None:
                s += l1.val
                l1 = l1.next

            if l2 is not None:
                s += l2.val
                l2 = l2.next

            carry = s // 10
            cur.next = ListNode(s % 10)
            cur = cur.next

        return dummy.next

File: Medium/0002-add-two-numbers/cpp/addtwonumber.cpp
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x = 0) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // Dummy node to simplify head handling.
        ListNode dummy(0);
        ListNode* cur = &dummy;

        int carry = 0;

        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            int sum = carry;

            if (l1 != nullptr) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr) {
                sum += l2->val;
                l2 = l2->next;
            }

            carry = sum / 10;

            // Create new digit node and move forward.
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
        }

        return dummy.next;
    }
};
